components/log/log.h

```diff
@@ -0,0 +1,113 @@
+#ifndef DATADOG_PHP_PROFILER_LOG_H
+#define DATADOG_PHP_PROFILER_LOG_H
+
+#include <pthread.h>
+#include <stdint.h>
+#include <string_view/string_view.h>
+
+#if __cplusplus
+#define C_STATIC(...)
+#else
+#define C_STATIC(...) static __VA_ARGS__
+#endif
+
+typedef enum {
+    DATADOG_PHP_LOG_UNKNOWN = -1,
+    DATADOG_PHP_LOG_OFF = 0,
+    DATADOG_PHP_LOG_ERROR,
+    DATADOG_PHP_LOG_WARN,
+    DATADOG_PHP_LOG_INFO,
+    DATADOG_PHP_LOG_DEBUG,
+} datadog_php_log_level;
+
+// This is for the compiler to use; please do not use it directly!
+typedef struct datadog_php_logger_s {
+    int descriptor;  // borrowed, not owned
+    int log_level;
+    pthread_mutex_t *mutex;  // borrowed, not owned
+} datadog_php_logger;
+
+#define DATADOG_PHP_LOGGER_INIT \
+    { -1, DATADOG_PHP_LOG_UNKNOWN, NULL }
+
+/**
+ * Returns true if the logger is considered to be valid by the implementation;
+ * returns false otherwise.
+ * @param logger
+ * @return
+ */
+bool datadog_php_logger_valid(datadog_php_logger *logger);
+
+/**
+ * Returns a newly constructed logger. The result may be valid or invalid; check
+ * with the function `datadog_php_logger_valid`.
+ *
+ * For now, only file descriptors are expected to work, but other types are
+ * expected to work in the future.
+ *
+ * @param descriptor
+ * @param log_level
+ * @return
+ */
+datadog_php_logger datadog_php_logger_ctor(int descriptor, datadog_php_log_level log_level, pthread_mutex_t *mutex);
+
+/**
+ * Sets the .descriptor to -1, .log_level to `DATADOG_PROFILER_LOG_UNKNOWN`,
+ * and the .mutex to NULL.
+ *
+ * This is not threadsafe! It also doesn't close the descriptor nor destroy the
+ * mutex, as they are borrowed, not owned. Only do this when the program is
+ * back in single-threaded mode, or you risk a race condition.
+ *
+ * @param logger
+ */
+void datadog_php_logger_dtor(datadog_php_logger *logger);
+
+/**
+ * If the `logger` is set to at least `level`, then the message will be logged.
+ * Otherwise it will be ignored.
+ * @param logger
+ * @param level
+ * @param message
+ */
+void datadog_php_log(datadog_php_logger *logger, datadog_php_log_level level, datadog_php_string_view message);
+
+/**
+ * If the `logger` is set to at least `level`, then the messages will be logged.
+ * Otherwise they will be ignored.
+ * @param logger
+ * @param level
+ * @param n_messages
+ * @param messages
+ * @return Bytes written
+ */
+int64_t datadog_php_logv(datadog_php_logger *logger, datadog_php_log_level level, size_t n_messages,
+                         datadog_php_string_view messages[C_STATIC(n_messages)]);
```

Comment:
I'm not sure what to make of this particular design choice to merely concatenate messages?
Is this a tactic to avoid constructing a big buffer?
I'd prefer a trivial printf-style functionality (maybe additionally). This also alleviates us the need from having to pre-format any non-string values (numbers ...) in case the log level is too low for the message to be logged at all.

Reply:
Almost all current messages are literals or literals plus a string generated by some other system (like `strerror`). This design avoids creating heap-allocated buffers just to call the log function (which may not log the message). So, yeah, this was basically the minimum design.

The design can improve, of course.

Reply:
Yeah, no blocker here, mainly wondering and suggesting.